estimagic/differentiation/numdiff_np.py

```diff
@@ -166,12 +180,162 @@ def _consolidate_one_step_derivatives(candidates, preference_order):
 def _consolidate_extrapolated(candidates):
     """Get the best possible derivative estimate, given an error estimate.
 
+    Going through ``candidates`` select the best derivative estimate element-wise using
+    the estimated candidates, where best is defined as minimizing the error estimate
+    from the Richardson extrapolation.
+
     See https://tinyurl.com/ubn3nv5 for corresponding code in numdifftools and
     https://tinyurl.com/snle7mb for an explanation of how errors of Richardson
     extrapolated derivative estimates can be estimated.
 
+    Args:
+        candidates (OrderedDict): Dictionary containing different derivative estimates
+            and their error estimates.
+
+    Returns:
+        consolidated (np.ndarray): Array of same shape as input derivative estimates.
+
     """
-    raise NotImplementedError
+    # first find minimum over steps for each method
+    candidate_derivatives = OrderedDict()
+    candidate_errors = OrderedDict()
+
+    for key in candidates.keys():
+        _limit = candidates[key]["derivative"]
+        _error = candidates[key]["error"]
+
+        derivative, error = _get_best_estimate_single_method(_limit, _error)
+
+        candidate_derivatives[key] = derivative
+        candidate_errors[key] = error
+
+    # second find minimum over methods
+    consolidated = _get_best_estimate_along_methods(
+        candidate_derivatives, candidate_errors
+    )
+    return consolidated
+
+
+def _compute_richardson_candidates(jac_candidates, steps, n_steps):
+    """Compute derivative candidates using Richardson extrapolation.
+
+    Args:
+        jac_candidates (dict): Dictionary with (traditional) derivative estimates from
+            different methods.
+        steps (namedtuple): Namedtuple with the field names pos and neg. Each field
+            contains a numpy array of shape (n_steps, len(x)) with the steps in
+            the corresponding direction. The steps are always symmetric, in the sense
+            that steps.neg[i, j] = - steps.pos[i, j] unless one of them is NaN.
+        n_steps (int): Number of steps needed. For central methods, this is
+            the number of steps per direction. It is 1 if no Richardson extrapolation
+            is used.
+
+    Returns:
+        richardson_candidates (OrderedDict): Dictionary with derivative estimates and
+            error estimates from different methods.
+    """
+    richardson_candidates = OrderedDict()
+    for method in ["forward", "backward", "central"]:
+        for num_terms in range(1, n_steps):
+            derivative, error = richardson_extrapolation(
+                jac_candidates[method], steps, method, num_terms
+            )
+            richardson_candidates[method + str(num_terms)] = {
```

Comment:
I have the feeling that an OrderedDict is not the best data structure to represent this. Since the keys are a composite of the metho and `str(num_terms)`, wouldn't it be better to have a dictionary of list, where the keys are just the methods and the values are lists with one entry per num_terms. The drawback is that num_terms starts at one and list indices at zero, but I still find it better. 

If not, we still don't need an explicit ordered dict for what you are doing. Dicts are ordered in cpython since 3.6 and in all other python implementations since 3.7. I think the estimagic user-base that uses a fancy 3.6 Python implementation is small enough to neglect them.

Reply:
Ok, you convinced me here.